# Big Data Analytics and Distributed ML Inference with Apache Spark

## 📌 Project Overview

This project demonstrates distributed Big Data analytics and Machine Learning inference workflows using Apache Spark.

The project focuses on processing large-scale web server search term data with Spark DataFrames and executing distributed predictive analytics using pretrained Machine Learning models.

The workflow combines:
- Spark-based data processing
- analytical query execution
- distributed ML model inference
- predictive sales forecasting
- scalable Big Data processing techniques

The project simulates a modern enterprise Big Data analytics environment where Apache Spark is used for both large-scale analytical processing and distributed Machine Learning execution.

---

## 🎯 Objectives

- Process structured datasets using Apache Spark
- Perform analytical operations with Spark DataFrames
- Analyze web server search behavior
- Deploy pretrained Machine Learning models
- Execute distributed ML inference workflows
- Generate future sales forecasting predictions
- Demonstrate scalable Big Data analytics architecture

---

## 🛠 Technologies Used

- Apache Spark
- PySpark
- Spark DataFrames
- Machine Learning Inference
- Predictive Analytics
- Distributed Processing
- Python
- Jupyter Notebook

---

## 🏗 Big Data Architecture

The project workflow includes:

- ingestion of CSV-based web server datasets
- distributed processing using Spark DataFrames
- analytical aggregation and filtering operations
- loading pretrained ML models
- distributed prediction execution in Spark
- forecasting future sales trends

The architecture demonstrates how Apache Spark can be used for both analytical engineering and scalable Machine Learning inference workflows.

## 1. Spark Environment Setup

This step prepares the Apache Spark environment for distributed data processing and machine learning inference.

The required PySpark libraries are installed, and a Spark session is initialized to enable DataFrame operations and ML workflows.

In [32]:
# Install required Spark libraries
!pip install pyspark
!pip install findspark

### Initialize Spark

The Spark environment is initialized using `findspark`, followed by importing the required Spark modules.

In [33]:
import findspark
findspark.init()

from pyspark import SparkContext, SparkConf
from pyspark.sql import SparkSession

### Create Spark Session

A Spark session is created as the main entry point for working with Spark DataFrames and ML operations.

In [34]:
# Create Spark session
spark = SparkSession \
    .builder \
    .appName("Big Data Analytics and Distributed ML Inference") \
    .getOrCreate()

sc = spark.sparkContext

## 2. Dataset Loading into Spark DataFrame

The web server search term dataset is loaded from the local project repository into a Spark DataFrame for distributed analytical processing.

The dataset contains search activity information collected from an e-commerce platform and is used to analyze search behavior and query frequency.

In [36]:
# Load local CSV dataset into Spark DataFrame
df = spark.read.csv(
    "../datasets/searchterms.csv",
    header=True,
    inferSchema=True
)

### Load Dataset into Spark DataFrame

The CSV dataset is loaded into a Spark DataFrame using automatic schema inference and header detection.

### Dataset Validation

The dataset structure and size are validated by checking the number of rows, columns, and previewing sample records.

In [37]:
# Print dataset dimensions
print("Rows:", df.count())
print("Columns:", len(df.columns))

Rows: 10000
Columns: 4


### Validation Results

The dataset was successfully loaded into a Spark DataFrame.

Validation results:
- total rows: 10,000
- total columns: 4
- the dataset was loaded correctly without schema or ingestion errors

The validation process confirms that the dataset is ready for distributed analytical processing.

## 3. Spark DataFrame Exploration

The Spark DataFrame schema and dataset structure are explored to validate column types and prepare the dataset for analytical operations.

In [16]:
# Display DataFrame schema
df.printSchema()

root
 |-- day: integer (nullable = true)
 |-- month: integer (nullable = true)
 |-- year: integer (nullable = true)
 |-- searchterm: string (nullable = true)



### Exploration Results

The Spark DataFrame schema was successfully analyzed.

Schema observations:
- `day`, `month`, and `year` were automatically inferred as integer fields
- `searchterm` was detected as a string column
- Spark schema inference correctly identified structured data types
- the dataset structure is suitable for analytical aggregation and ML processing workflows

The DataFrame is now fully prepared for search analytics and distributed Spark operations.

## 4. Search Term Analytics

Distributed analytical queries are executed on the Spark DataFrame to identify search behavior patterns and analyze the frequency of search terms within the e-commerce dataset.

### Search Frequency Analysis

The dataset is filtered to determine how many times a specific search term was queried by users.

In [17]:
from pyspark.sql.functions import count

df.filter(df.searchterm == "gaming laptop").count()

499

### Search Query Results

The term `gaming laptop` appeared 499 times in the dataset.

This indicates that gaming-related product searches represent a significant portion of user search activity within the analyzed e-commerce platform.

### Top Search Terms Analysis

Aggregation operations are executed using Spark DataFrames to identify the most frequently searched product terms.

In [26]:
from pyspark.sql.functions import count

df.groupby("searchterm") \
    .agg(count("*").alias("count")) \
    .orderBy("count", ascending=False) \
    .show(5)

[Stage 22:===========================================>         (166 + 10) / 200]

+-------------+-----+
|   searchterm|count|
+-------------+-----+
|mobile 6 inch| 2312|
|    mobile 5g| 2301|
|mobile latest| 1327|
|       laptop|  935|
|  tablet wifi|  896|
+-------------+-----+
only showing top 5 rows



### Analytical Results

The analysis identified the most frequently searched product categories and device types.

Key observations:
- mobile device searches dominate the dataset
- `mobile 6 inch` and `mobile 5g` are the most frequently searched terms
- smartphone-related searches appear significantly more often than laptop or tablet searches
- search behavior indicates strong user interest in mobile technology products

The Spark aggregation workflow successfully demonstrated distributed analytical processing using Spark DataFrames.

In [38]:
from pyspark.ml.regression import LinearRegressionModel

# Load pretrained SparkML model from local repository
model = LinearRegressionModel.load(
    "../models/sales_prediction.model"
)

## Model Deployment Results

The pretrained SparkML regression model was successfully loaded and deployed inside the Apache Spark environment.

Key observations:

- the regression model was loaded into the Spark environment
- distributed ML inference is now available
- the environment is prepared for predictive analytics execution
- Spark ML successfully initialized the forecasting workflow
- reusable model deployment workflow was validated successfully

The deployment process demonstrates scalable Machine Learning model integration and distributed inference execution using Apache Spark.

## 6. Distributed ML Inference and Sales Forecasting

The pretrained Spark ML regression model is used to generate distributed sales forecasting predictions for future periods.

A Spark DataFrame containing future year values is created and transformed into feature vectors for distributed ML inference execution.

### Prepare Forecasting Dataset

A prediction dataset is created for future sales forecasting using the year 2023 as input for the regression model.

In [22]:
from pyspark.ml.feature import VectorAssembler

# Create forecasting dataset
data = [[2023]]
columns = ["year"]

df_2023 = spark.createDataFrame(data, columns)

# Transform dataset into feature vectors
assembler = VectorAssembler(
    inputCols=["year"],
    outputCol="features"
)

final_data = assembler.transform(df_2023)

### Feature Engineering Results

The forecasting dataset was successfully transformed into Spark ML feature vectors.

Key observations:
- the future prediction dataset contains year-based forecasting input
- Spark ML feature engineering was successfully executed
- the dataset is now prepared for distributed prediction processing

### Execute Distributed Predictions

The pretrained Spark ML model is executed on the forecasting dataset to generate future sales predictions.

In [24]:
# Execute distributed predictions
predictions = model.transform(final_data)

# Display prediction results
predictions.select("prediction").show()

+------------------+
|        prediction|
+------------------+
|175.16564294006457|
+------------------+



### Forecasting Results

The distributed Spark ML inference workflow successfully generated future sales forecasting predictions.

Prediction output:
- forecasted sales value for 2023: approximately 175.17

Key observations:
- the pretrained regression model successfully executed distributed inference
- Apache Spark handled predictive analytics processing within the distributed environment
- the forecasting workflow demonstrated scalable ML execution using Spark ML

The project successfully combined Big Data analytics and Machine Learning inference within a unified Apache Spark architecture.

## 7. Conclusion

This project demonstrated enterprise-style Big Data analytics and distributed Machine Learning inference workflows using Apache Spark.

Implemented capabilities:
- distributed CSV data processing
- Spark DataFrame analytics
- analytical aggregation workflows
- ML model deployment
- distributed Spark ML inference
- predictive sales forecasting

The project successfully demonstrated how Apache Spark can be used as a unified platform for scalable analytical engineering and Machine Learning execution.